lib/postgrex/extensions/hstore.ex

```diff
@@ -38,22 +40,40 @@ defmodule Postgrex.Extensions.HStore do
     <<value_byte_size::int32>> <> value
   end
 
-  def decode_hstore(<<_length::int32, pairs::binary>>) do
-    decode_hstore_payload(pairs, %{})
+  def decode_hstore(<<_length::int32, pairs::binary>>, :reference) do
+    decode_hstore_ref(pairs, %{})
+  end
+  def decode_hstore(<<_length::int32, pairs::binary>>, :copy) do
+    decode_hstore_copy(pairs, %{})
+  end
+
+  defp decode_hstore_ref(<<>>, acc) do
+    acc
+  end
+
+  # in the case of a NULL value, there won't be a length
+  defp decode_hstore_ref(<<key_length::int32, key::binary(key_length),
+                                 -1::int32, rest::binary>>, acc) do
+    decode_hstore_ref(rest, Map.put(acc, key, nil))
+  end
+
+  defp decode_hstore_ref(<<key_length::int32, key::binary(key_length),
+                        value_length::int32, value::binary(value_length), rest::binary>>, acc) do
+    decode_hstore_ref(rest, Map.put(acc, key, value))
   end
 
-  defp decode_hstore_payload(<<>>, acc) do
+  defp decode_hstore_copy(<<>>, acc) do
     acc
   end
 
   # in the case of a NULL value, there won't be a length
-  defp decode_hstore_payload(<<key_length::int32, key::binary(key_length),
-                             -1::int32, rest::binary>>, acc) do
-    decode_hstore_payload(rest, Map.put(acc, key, nil))
+  defp decode_hstore_copy(<<key_length::int32, key::binary(key_length),
+                                 -1::int32, rest::binary>>, acc) do
+    decode_hstore_copy(rest, Map.put(acc, :binary.copy(key), nil))
   end
 
-  defp decode_hstore_payload(<<key_length::int32, key::binary(key_length),
+  defp decode_hstore_copy(<<key_length::int32, key::binary(key_length),
                         value_length::int32, value::binary(value_length), rest::binary>>, acc) do
-    decode_hstore_payload(rest, Map.put(acc, key, value))
+    decode_hstore_copy(rest, Map.put(acc, :binary.copy(key), :binary.copy(value)))
```

Comment:
@ericmj can you expand? Would you prefer the copy to occur on the binary of the whole map?